keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1392 @@
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+None so far
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+SLOs are not applicable to pod status fields. Overall Kubernetes node level SLOs
+may leverage this feature.
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+The main drawback associated with the new pod sandbox conditions involves a
+slight potential increase in calls to the API Server from Kubelet to patch
+`status` = `true` for the new conditions in a pod's status. Typically, this
+would involve two extra patch calls for pod status in the lifetime of most pods
+(if the status manager does not batch them with other pod status updates): one
+when pod sandbox creation actually starts and the other when pod sandbox
+creation completes successfully. However, there could be a higher number of
+patch calls to API Server if the pod sandbox environment (like a microvm) starts
+successfully and then crashes in a re-start loop.
+
+Caching of updates to pod status by the pod status manager and batching pod
+status updates (which is already in place) can help mitigate frequent patch
+calls to API server.
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Dedicated fields or annotations for the pod sandbox creation timestamps
+
+Timestamps around start and completion of pod sandbox creation may be surfaced
+as dedicated fields in the pod status rather than pod conditions. However, since
+the start and creation of pod sandbox are essentially "milestones" in the life
+of a pod (similar to Scheduled, Ready, etc), pod conditions is the ideal place
+to surface these.
+
+Dedicated annotations on the pod for surfacing this data is another potential
+approach. However, usage of annotations for Kubelet managed data is typically
+discouraged.
+
+### Surface pod sandbox creation latency instead of timestamps
```

Comment:
What about just surfacing the creation completion time? Does the start time or diff here add much on top of the scheduled and/or initialized time?

Reply:
The pod sandbox can get re-created multiple times during the lifetime of a pod (as mentioned [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R503)). Every time the pod sandbox is re-created, the sandbox completion time would be updated. So to calculate the latency of the sandbox creation process, we need the sandbox creation start time. The Scheduled condition could be a long way in the past and not relevant in the context of a sandbox re-creation scenario for a pod. The Initialized condition will get updated after sandbox creation succeeds if init containers are present and therefore too late in the context of sandbox creation latency.

Reply:
Explained [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R225)